### PR TITLE
reporting cgroups limit exceedings into diag_messages job resource

### PR DESCRIPTION
**Pull request by vchlum**
#### Describe Bug or Feature

I would like to offer you an improvement of cgroups hook. In this change, a new resource `diag_messages` is created by cgroups hook. This resource is used to inform the user that the jobs' limits were exceeded on nodes. As the resource is shown in qstat, the user can easily see it. It can also be shown in a custom GUI app etc... This resource diag_messages can be also used for other purposes in the future.

#### Describe Your Change

The cgroups hook creates a new resource `diag_messages` on vnode creating. As only 'used' job resources can be changed from periodic hooks, this seems to be the convenient way. The cgroups periodic hook will write a msg that the limits were exceeded into this 'used' resource. Sister moms messages are merged by openpbs using JSON into one message. Hence the msg format is JSON. Also, more than one limit can be exceeded on a node - the messages of the node are concated. 

Resource value example for a job with exceeding limits on two moms:
```
resources_used.diag_messages = '{"torque2": "Cgroup memsw limit exceeded", "torque1": "Cgroup memsw limit exceeded"}'

```
I realize this change uses the 'used' resource in an alternative/creative way. I understand if you do not want to merge.

#### Attach Test and Valgrind Logs/Output

[PTL_diag_messages.txt](https://github.com/user-attachments/files/16051521/PTL_diag_messages.txt)

### CODE REVIEW

File: src/hooks/cgroups/pbs_cgroups.PY [hunk 1]
```diff
@@ -3041,6 +3043,33 @@ class CgroupUtils(object):
...
+        if 'diag_messages' not in resc_used.keys():
+            resc_used['diag_messages'] = json.dumps(json_msg)
+        elif concate:
+            try:
+                json_resc = json.loads(resc_used['diag_messages'])
+            except Exception:
+                json_resc = {}
+            for k in json_msg.keys():
+                if k in json_resc.keys() and not json_msg[k] in json_resc[k]:
+                    json_resc[k] += ', %s' % json_msg[k]
```
Comment (bhroam) [inline src/hooks/cgroups/pbs_cgroups.PY:3066]: I'm not sure about OpenPBS, but at least with commercial, we no longer support python releases older than 3.6.  A new type of string formatting was introduced in 3.6 called fstrings.  They are much clearer and easier to use than the old string formatting.

f"foo: {foo} bar: {bar} baz: {baz[5]}"
Anything between the {} is python code to be evaluated.  Note that when using a dictionary, you need to use alternate quotes f"{foo['bar']}"

File: src/hooks/cgroups/pbs_cgroups.PY [hunk 2]
```diff
@@ -4932,6 +4961,11 @@ class CgroupUtils(object):
             pbs.logmsg(pbs.EVENT_DEBUG4, '%s: Job %s is not running' %
                        (caller_name(), jobid))
             return
+        if pbs.event().type == pbs.EXECHOST_PERIODIC or \
+           pbs.event().type == pbs.EXECJOB_EPILOGUE:
```
Comment (bhroam) [inline src/hooks/cgroups/pbs_cgroups.PY:4965]: Consider: if pbs.event().type in [pbs.EXECHOST_PERIODIC, pbs.EXECJOB_EPILOGUE]: 

here and below

File: test/tests/functional/pbs_cgroups_hook.py [hunk 3]
```diff
@@ -2741,6 +2741,52 @@ def test_cgroup_enforce_memsw(self):
...
+        fn = self.paths[self.hosts_list[0]]['memory']
+        fn = os.path.join(fn, 'memory.memsw.limit_in_bytes')
+        if not self.is_file(fn, self.hosts_list[0]):
+            self.skipTest('vmem resource not present on node')
+
+        # Make sure job history is enabled to see when job is gone
+        a = {'job_history_enable': 'True'}
+        rc = self.server.manager(MGR_CMD_SET, SERVER, a)
+        self.assertEqual(rc, 0)
+        self.server.expect(SERVER, {'job_history_enable': 'True'})
```
Comment (bhroam) [inline test/tests/functional/pbs_cgroups_hook.py:2768]: You probably don't need this.  You asserted that the manager() call succeeded.

File: test/tests/functional/pbs_cgroups_hook.py [hunk 4]
```diff
@@ -2741,6 +2741,52 @@ def test_cgroup_enforce_memsw(self):
...
+                                      self.swapctl, ''))
+
+        a = {
+            'Resource_List.select':
+            '1:ncpus=1:mem=400mb:vmem=420mb:host=%s' % self.hosts_list[0]}
+        j = Job(TEST_USER, attrs=a)
+        j.create_script(self.eatmem_job1)
+        jid = self.server.submit(j)
+        a = {'job_state': 'F'}
+        self.server.expect(JOB, a, jid, extend='x')
```
Comment (bhroam) [inline test/tests/functional/pbs_cgroups_hook.py:2780]: Consider using an offset if you have an idea of how long the job takes.

File: test/tests/functional/pbs_cgroups_hook.py [hunk 5]
```diff
@@ -2741,6 +2741,52 @@ def test_cgroup_enforce_memsw(self):
...
+        resc = ['resources_used.diag_messages']
+        s = self.server.status(JOB, resc, id=jid, extend='x')
+        dmsg = s[0]['resources_used.diag_messages'].replace("'", "")
+        self.logger.info(dmsg)
```
Comment (bhroam) [inline test/tests/functional/pbs_cgroups_hook.py:2784]: How about f"diag_message: {dmsg}"
Just printing a variable out is confusing.
If this is a dev debug message, consider removing it.

**Review by bhroam — CHANGES_REQUESTED** [review]

File: src/hooks/cgroups/pbs_cgroups.PY (hunk 1 above)
Comment (vchlum) [reply src/hooks/cgroups/pbs_cgroups.PY]: You are right. It is nicer. Moreover, we already use it in PTL. Changed.

File: test/tests/functional/pbs_cgroups_hook.py (hunk 5 above)
Comment (vchlum) [reply test/tests/functional/pbs_cgroups_hook.py]: You are right, it was a debug msg. Thank you for pointing out. Removed.

**Comment by vchlum** [comment]
Thank you @bhroam for your comments. I addressed all of them.

**Review by bhroam — APPROVED** [review]

**Review by bayucan — APPROVED** [review]
Looks good.

**Pull request merged**